File: Problem_List/00140.py
class Solution:
    def wordBreak(self, s: str, wordDict: list[str]) -> list[str]:
        ans = []
        wordDict = set(wordDict)
        def rec(s,tmp_str):
            if len(s) == 0:
                ans.append(tmp_str[:-1])
            tmp_s = ""
            for i,char in enumerate(s):
                tmp_s += char
                if tmp_s in wordDict:
                    rec(s[i+1:],tmp_str+tmp_s+' ')
            return
        rec(s,'')
        return ans

class Solution:
    def wordBreak(self, s: str, wordDict: list[str]) -> list[str]:
        ans = []
        from collections import defaultdict
        words = defaultdict(lambda:[])

        for word in wordDict:
            words[word[0]].append(word)

        def rec(start_idx,tmp_str):
            if start_idx == len(s):
                ans.append(tmp_str[:-1])
                return
            
            for word in words[s[start_idx]]:
                if s[start_idx:start_idx+len(word)] == word:
                    rec(start_idx+len(word),tmp_str+word+' ')
        rec(0,'')
        return ans

File: ProblemList/00140.cpp
#include <vector>
#include <string>
using namespace std;

// using trie
struct TrieNode{
    vector<TrieNode*> next;
    bool end;
    TrieNode(): next(26,nullptr), end(false) {};
};
class Solution {
private:
    vector<string> ans;
    TrieNode* trieRoot;
    void rec(int idx, TrieNode* curNode, string& curS, string& s){
        if(idx == s.size()){
            if(curNode -> end) ans.push_back(curS);
            return;
        } 

        if(curNode -> end){
            string newS = curS + " ";
            rec(idx,trieRoot,newS,s);
        }

        if(curNode->next[s[idx]-'a']){ 
            curS += s[idx];
            rec(idx+1,curNode->next[s[idx]-'a'],curS,s);
            curS.pop_back();
        }
        return;
    }
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        trieRoot = new TrieNode();
        for(auto& word: wordDict){
            TrieNode* cur = trieRoot;
            for(auto ch: word){
                if(cur -> next[ch-'a'] == nullptr){
                    TrieNode* newNode = new TrieNode();
                    cur -> next[ch-'a'] = newNode;
                }
                cur = cur -> next[ch-'a'];
            }
            cur -> end = true;
        }
        
        string cur;
        rec(0,trieRoot,cur,s);
        return ans;
    }
};